File: dotnet/base/Mcma.Data/DocumentDatabase/DocumentDatabaseMutex.cs

using System;
using System.Threading.Tasks;
using Mcma.Logging;

namespace Mcma.Data
{
    public abstract class DocumentDatabaseMutex : IDocumentDatabaseMutex
    {
        protected DocumentDatabaseMutex(string mutexName, string mutexHolder, TimeSpan? lockTimeout, ILogger logger = null)
        {
            MutexName = mutexName ?? throw new ArgumentNullException(nameof(mutexName));
            MutexHolder = mutexHolder ?? throw new ArgumentNullException(nameof(mutexHolder));
            LockTimeout = lockTimeout ?? TimeSpan.FromMinutes(1);
            Logger = logger;
        }

        protected string MutexName { get; }
        
        protected string MutexHolder { get; }
        
        protected TimeSpan LockTimeout { get; }
        
        protected ILogger Logger { get; }
        
        protected bool HasLock { get; set; }
        
        protected abstract string VersionId { get; }
        
        public async Task LockAsync()
        {
            if (HasLock)
                throw new McmaException("Cannot lock when already locked");

            Logger?.Debug("Requesting lock for mutex '" + MutexName + "' by '" + MutexHolder + "'");
            while (!HasLock) {
                try
                {
                    await PutLockDataAsync();
                    var lockData = await GetLockDataAsync();
                    HasLock = lockData?.MutexHolder == MutexHolder && lockData?.VersionId == VersionId;
                }
                catch
                {
                    var lockData = await GetLockDataAsync();
                    if (lockData != null)
                    {
                        if (lockData.Timestamp < DateTime.UtcNow - LockTimeout)
                        {
                            Logger?.Warn("Deleting stale lock for mutex '" + MutexName + "' by '" + lockData.MutexHolder + "'");
                            try
                            {
                                await DeleteLockDataAsync(lockData.VersionId);
                            }
                            catch
                            {
                                // nothing else we can do at this point
                            }
                        }
                    }
                }

                if (!HasLock) {
                    await Task.Delay(TimeSpan.FromSeconds(0.5));
                }
            }
            Logger?.Debug("Acquired lock for mutex '" + MutexName + "' by '" + MutexHolder + "'");
        }

        public async Task UnlockAsync()
        {
            if (!HasLock) {
                throw new McmaException("Cannot unlock when not locked");
            }

            await DeleteLockDataAsync(VersionId);
            HasLock = false;
            Logger?.Debug("Released lock for mutex '" + MutexName + "' by '" + MutexHolder + "'");
        }

        protected abstract Task PutLockDataAsync();

        protected abstract Task<LockData> GetLockDataAsync();

        protected abstract Task DeleteLockDataAsync(string versionId);
    }
}

File: base/data/lib/document-database/document-database-mutex.ts
import { Logger, McmaException, Utils } from "@mcma/core";

export interface DocumentDatabaseMutex {
    lock(): Promise<void>;
    tryLock(): Promise<boolean>;
    unlock(): Promise<void>;
}

export interface LockData {
    mutexHolder: string;
    timestamp: number;
    versionId?: string;
}

export abstract class DocumentDatabaseMutex {
    protected hasLock = false;

    protected constructor(
        public readonly mutexName: string,
        protected readonly mutexHolder: string,
        protected readonly lockTimeout = 300000,
        protected readonly logger?: Logger
    ) {
    }

    protected abstract get versionId(): string;

    protected abstract putLockData(): Promise<void>;

    protected abstract getLockData(): Promise<LockData>;

    protected abstract deleteLockData(versionId: string): Promise<void>;

    async lock() {
        if (this.hasLock) {
            throw new McmaException("Cannot lock when already locked");
        }

        this.logger?.debug("Requesting lock for mutex '" + this.mutexName + "' by '" + this.mutexHolder + "'");
        while (!this.hasLock) {
            try {
                await this.putLockData();
                const lockData = await this.getLockData();
                this.hasLock = lockData?.mutexHolder === this.mutexHolder && lockData?.versionId === this.versionId;
            } catch (error) {
                const lockData = await this.getLockData();
                if (lockData) {
                    if (lockData.timestamp < Date.now() - this.lockTimeout) {
                        this.logger?.warn("Deleting stale lock for mutex '" + this.mutexName + "' by '" + lockData.mutexHolder + "'");
                        try {
                            await this.deleteLockData(lockData.versionId);
                        } catch (error) {
                        }
                    }
                }
            }

            if (!this.hasLock) {
                await Utils.sleep(500);
            }
        }
        this.logger?.debug("Acquired lock for mutex '" + this.mutexName + "' by '" + this.mutexHolder + "'");
    }

    async tryLock(): Promise<boolean> {
        if (this.hasLock) {
            throw new McmaException("Cannot lock when already locked");
        }

        let tryAgain: boolean;

        this.logger?.debug("Requesting lock for mutex '" + this.mutexName + "' by '" + this.mutexHolder + "'");
        do {
            tryAgain = false;
            try {
                await this.putLockData();
                const lockData = await this.getLockData();
                this.hasLock = lockData?.mutexHolder === this.mutexHolder && lockData?.versionId === this.versionId;
            } catch (error) {
                const lockData = await this.getLockData();
                if (lockData) {
                    if (lockData.timestamp < Date.now() - this.lockTimeout) {
                        this.logger?.warn("Deleting stale lock for mutex '" + this.mutexName + "' by '" + lockData.mutexHolder + "'");
                        try {
                            await this.deleteLockData(lockData.versionId);
                        } catch (error) {
                        }
                        tryAgain = true;
                    }
                }
            }

            if (tryAgain) {
                await Utils.sleep(500);
            }
        } while (tryAgain);

        if (this.hasLock) {
            this.logger?.debug("Acquired lock for mutex '" + this.mutexName + "' by '" + this.mutexHolder + "'");
        } else {
            this.logger?.debug("Failed to acquire lock for mutex '" + this.mutexName + "' by '" + this.mutexHolder + "'");
        }

        return this.hasLock;
    }

    async unlock() {
        if (!this.hasLock) {
            throw new McmaException("Cannot unlock when not locked");
        }

        await this.deleteLockData(this.versionId);
        this.hasLock = false;
        this.logger?.debug("Released lock for mutex '" + this.mutexName + "' by '" + this.mutexHolder + "'");
    }
}
